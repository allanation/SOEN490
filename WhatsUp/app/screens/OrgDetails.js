/* eslint-disable no-unused-vars */
import {
  StyleSheet,
  View,
  ScrollView,
  Platform,
  Text,
  Dimensions,
  Alert,
} from "react-native";
import React, { useState } from "react";
import Screen from "../components/Screen";
import MapView, { PROVIDER_GOOGLE, Marker } from "react-native-maps";
import TitleHeaders from "../components/TitleHeaders";
import colors from "../config/colors";
import { Ionicons } from "@expo/vector-icons";
import PropTypes from "prop-types";
import EventTagsList from "../components/EventTagsList";
import { convertStartDate } from "./AttendeeDashboard.js";
import { format } from "date-fns";
import { useNavigation } from "@react-navigation/native";
import { Storage } from "expo-storage";
import { db } from "../firebase";
import {
  collection,
  getDocs,
  query,
  where,
  deleteDoc,
} from "firebase/firestore";

OrgDetails.propTypes = {
  route: PropTypes.any,
};

export const convertTime = (number) => {
  return number ? format(new Date(number), "hh:mm aaaaa'm'") : "";
};

export const storeNewEvent = async (newEvent) => {
  try {
    const jsonValue = JSON.stringify(newEvent);
    await Storage.setItem({
      key: "newEvent",
      value: jsonValue,
    });
  } catch (e) {
    console.log(e);
  }
};

export const storePOC = async (POC) => {
  try {
    const jsonValue = JSON.stringify(POC);
    await Storage.setItem({
      key: "POC",
      value: jsonValue,
    });
  } catch (e) {
    console.log(e);
  }
};

export const storeDates = async (eventDates) => {
  try {
    const jsonValue = JSON.stringify(eventDates);
    await Storage.setItem({
      key: "eventDates",
      value: jsonValue,
    });
  } catch (e) {
    console.log(e);
  }
};

export const storeDay = async (days) => {
  try {
    await Storage.setItem({
      key: "days",
      value: days,
    });
  } catch (e) {
    console.log(e);
  }
};

export const storeItinerary = async (itinerary) => {
  try {
    const jsonValue = JSON.stringify(itinerary);
    await Storage.setItem({
      key: "itinerary",
      value: jsonValue,
    });
  } catch (e) {
    console.log(e);
  }
};

export const storeTags = async (tags) => {
  try {
    const jsonValue = JSON.stringify(tags);
    await Storage.setItem({
      key: "tags",
      value: jsonValue,
    });
  } catch (e) {
    console.log(e);
  }
};

const onRemove = () => () => {};

function OrgDetails({ route }) {
  const tokyoRegion = {
    latitude: 35.6762,
    longitude: 139.6503,
    latitudeDelta: 0.01,
    longitudeDelta: 0.01,
  };
  const { prop } = route.params;
  const navigation = useNavigation();
  const [eventName, setEventName] = useState(prop.eventName);
  const [orgName, setOrgName] = useState(prop.orgName);
  const [location, setLocation] = useState(prop.location);
  const [link, setLink] = useState(prop.link);
  const [description, setDescription] = useState(prop.description);
  const [eventGuid, setEventGuid] = useState(prop.guid);
  const [coverImageName, setCoverImageName] = useState(prop.coverImage);
  const newEvent = {
    eventName: eventName,
    orgName: orgName,
    location: location,
    description: description,
    link: link,
    coverImage: coverImageName,
    guid: eventGuid,
  };

  const deleteEvent = async (eventGuid) => {
    const q = query(collection(db, "events"), where("guid", "==", eventGuid));
    const querySnapshot = await getDocs(q);
    if (!querySnapshot.empty) {
      querySnapshot.forEach((doc) => {
        try {
          deleteDoc(doc.ref);
          Alert.alert("Event was successfully deleted.");
          navigation.navigate("OrganizerDashboard");
        } catch (e) {
          console.log(e);
        }
      });
    }
  };

  //Point of contact details
  const [pocName, setPocName] = useState(prop.pocName);
  const [pocPhoneNum, setPocPhoneNum] = useState(prop.pocPhoneNum);
  const [pocEmail, setPocEmail] = useState(prop.pocEmail);
  const POC = {
    pocName: pocName,
    pocPhoneNum: pocPhoneNum,
    pocEmail: pocEmail,
  };

  //DateInfo details
  const [startDate, setStartDate] = useState({
    date: new Date(Date.now() + 24 * 60 * 60 * 1000),
    timestamp: prop.startDate,
  });
  const [endDate, setEndDate] = useState({
    date: new Date(),
    timestamp: prop.endDate,
  });
  const [startTime, setStartTime] = useState({
    date: new Date(),
    timestamp: prop.startTime,
  });
  const [endTime, setEndTime] = useState({
    date: new Date(),
    timestamp: prop.endTime,
  });

  const eventDates = {
    startDate: startDate.timestamp,
    startTime: startTime.timestamp,
    endDate: endDate.timestamp,
    endTime: endTime.timestamp,
  };

  //EventDays useState
  const [days, setDays] = useState(prop.days);

  //Itinerary details
  const [itinerary, setItinerary] = useState(prop.itinerary);

  //EventTags useState
  const [tags, setTags] = useState(prop.tags);

  return (
    <Screen style={{ padding: "5%", backgroundColor: "white" }}>
      <ScrollView style={{ width: "100%", display: "flex" }}>
        <View style={{ flexDirection: "row" }}>
          <TitleHeaders title={prop.eventName} />
          <View style={{ flexDirection: "row", marginLeft: "auto" }}>
            <Ionicons
              name="trash"
              onPress={() =>
                Alert.alert(
                  "Warning",
                  "Your event will be permanently deleted. Are you sure you wish to proceed?",
                  [
                    {
                      text: "Cancel",
                      style: "cancel",
                    },
                    {
                      text: "Delete Event",
                      onPress: () => deleteEvent(prop.guid),
                    },
                  ]
                )
              }
              size={20}
              color={colors.primary}
              style={{
                marginBottom: "1.5%",
                alignSelf: "flex-end",
                paddingRight: "5%",
              }}
            />
            <Ionicons
              name="pencil"
              testID="edit-event"
              onPress={() => {
                storeNewEvent(newEvent);
                storePOC(POC);
                storeDates(eventDates);
                storeDay(days);
                storeItinerary(itinerary);
                storeTags(tags);
                navigation.navigate("OrgReviewEvent");
              }}
              size={20}
              color={colors.primary}
              style={{
                marginBottom: "1.5%",
                alignSelf: "flex-end",
              }}
            />
          </View>
        </View>
        <Text
          style={{
            marginLeft: "1%",
            marginBottom: 5,
            color: "silver",
            marginTop: "1%",
          }}
        >
          By {prop.orgName}
        </Text>
        <View style={styles.iconText}>
          <Ionicons
            name="ios-location-outline"
            size={20}
            color={colors.primary}
          />
          <Text style={{ marginLeft: 10 }}>{prop.location}</Text>
        </View>
        <View style={styles.iconText}>
          <Ionicons
            name="ios-calendar-outline"
            size={20}
            color={colors.primary}
          />
          {convertStartDate(prop.startDate) ==
          convertStartDate(prop.endDate) ? (
            <Text style={{ marginLeft: 10 }}>{`${convertStartDate(
              prop.startDate
            )}`}</Text>
          ) : (
            <Text style={{ marginLeft: 10 }}>{`${convertStartDate(
              prop.startDate
            )} - ${convertStartDate(prop.endDate)}`}</Text>
          )}
        </View>
        <View style={styles.iconText}>
          <Ionicons name="ios-time-outline" size={20} color={colors.primary} />
          <Text style={{ marginLeft: 10 }}>{`${convertTime(
            prop.startTime
          )} - ${convertTime(prop.endTime)}`}</Text>
        </View>
        <Text style={styles.text}>Description</Text>
        <Text style={styles.description}>{prop.description}</Text>

        <Text style={styles.text}>Location</Text>
        <View style={{ borderRadius: 20, overflow: "hidden" }}>
          <MapView
            style={styles.map}
            initialRegion={tokyoRegion}
            provider={PROVIDER_GOOGLE}
          >
            <Marker coordinate={tokyoRegion} pinColor={colors.primary} />
          </MapView>
        </View>
        <View
          style={{
            marginTop: 10,
            flexDirection: "row",
            flexWrap: "wrap",
            width: "95%",
            alignSelf: "center",
            justifyContent: "flex-start",
          }}
        >
          {prop.tags.length > 0 ? (
            <View>
              <Text style={styles.text}>Tags</Text>
              <EventTagsList tags={prop.tags} onRemove={onRemove} />
            </View>
          ) : (
            <Text></Text>
          )}
        </View>
      </ScrollView>
    </Screen>
  );
}

const styles = StyleSheet.create({
  iconText: {
    flexDirection: "row",
    marginTop: 3,
    marginLeft: 8,
  },
  text: {
    color: "#100101",
    fontSize: 17,
    fontFamily: Platform.OS === "ios" ? "Helvetica Neue" : "sans-serif",
    marginTop: "5%",
    marginBottom: "2%",
    marginLeft: "1%",
  },
  description: {
    marginLeft: "1%",
    fontSize: 14,
    alignSelf: "flex-start",
    alignContent: "flex-start",
    fontFamily: Platform.OS === "ios" ? "Helvetica Neue" : "sans-serif",
  },
  map: {
    width: Dimensions.get("window").width,
    height: Dimensions.get("window").height / 4,
    borderRadius: 20,
    overflow: "hidden",
  },
});

export default OrgDetails;
