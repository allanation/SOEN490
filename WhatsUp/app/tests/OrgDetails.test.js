import "react-native";
import React from "react";
import { NavigationContainer } from "@react-navigation/native";
import {
  render,
  waitFor,
  fireEvent,
  screen,
} from "@testing-library/react-native";
import { Alert } from "react-native";
import OrgDetails from "../screens/OrgDetails";
import OrgReviewEvent from "../screens/OrgReviewEvent";

jest.useFakeTimers();                              
jest.mock('react-native-maps', () => {
  const React = jest.requireActual('react');
  const MapView = jest.requireActual('react-native-maps');
  
  class MockMapView extends React.Component {
    render() {
      return React.createElement('MapView', this.props, this.props.children);
    }
  }
  MockMapView.propTypes = MapView.propTypes;

  class MockCallout extends React.Component {
    render() {
      return React.createElement('Callout', this.props, this.props.children);
    }
  }
  MockCallout.propTypes = MapView.Callout.propTypes;
  MockMapView.Callout = MockCallout;

  class MockMarker extends React.Component {
    render() {
      return React.createElement('Marker', this.props, this.props.children);
    }
  }
  MockMarker.propTypes = MapView.Marker.propTypes;
  MockMapView.Marker = MockMarker;
  
  class MockPolyline extends React.Component {
    render() {
      return React.createElement('Polyline', this.props, this.props.children);
    }
  }
  MockPolyline.propTypes = MapView.Polyline.propTypes;
  MockMapView.Polyline = MockPolyline;
  /* eslint-enable react/prefer-stateless-function */

  return MockMapView;
});

const mockedParams = {
    route: { params: {prop: {eventName: 'Hi', orgName: 'Hi', location: 'Concordia', link: '', description: 'Fun', guid:'dfghjkl456789', coverImage: '',
    pocName: 'Andrea', pocPhoneNum: '514', pocEmail: 'a@e.com', startDate: 1687392780000, endDate: 1687392780000, endTime: 1679703182407, startTime: 1679703182407,
    days: 3, itinerary: [], tags: []} }},
    navigation: ''
    };

it("Renders correctly organizer view page", async () => {
    const tree = render(
      <NavigationContainer>
        <OrgDetails {...mockedParams} />
      </NavigationContainer>
    ).toJSON();
  
    expect(tree).toMatchSnapshot();
  });

  test("If click on delete icon, an alert should be prompted", async () => {
    render(
    <NavigationContainer>
        <OrgDetails {...mockedParams} />
    </NavigationContainer>
    );
    jest.spyOn(Alert, "alert");

    await waitFor(() => {
        fireEvent.press(screen.getByTestId("delete-event"));
        expect(Alert.alert).toHaveBeenCalled();
      });
    });

  test("If the edit icon is clicked, should navigate to Edit Event page", async () => {
    render(
    <NavigationContainer>
        <OrgDetails {...mockedParams} />
    </NavigationContainer>
    );

    await waitFor(() => {
        fireEvent.press(screen.getByTestId("edit-event"));
      });

      render(
              <NavigationContainer>
                <OrgReviewEvent {...mockedParams} />
              </NavigationContainer>
            );
      expect(screen.getByText("Edit Event")).toBeTruthy();
  });
