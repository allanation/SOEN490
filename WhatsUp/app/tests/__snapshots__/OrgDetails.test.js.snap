// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly organizer view page 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Array [
      Object {
        "flex": 1,
        "paddingTop": 20,
      },
      Object {
        "backgroundColor": "white",
        "padding": "5%",
      },
    ]
  }
>
  <Image
    source={
      Object {
        "uri": undefined,
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <RCTScrollView
    style={
      Object {
        "display": "flex",
        "width": "100%",
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Helvetica Neue",
                "marginLeft": "1%",
              },
              undefined,
              Object {
                "fontSize": 28,
                "fontWeight": "bold",
              },
            ]
          }
        >
          Hi
        </Text>
        <View
          style={
            Object {
              "flexDirection": "row",
              "marginLeft": "auto",
            }
          }
        >
          <Text />
          <Text />
        </View>
      </View>
      <Text
        style={
          Object {
            "color": "silver",
            "marginBottom": 5,
            "marginLeft": "1%",
            "marginTop": "1%",
          }
        }
      >
        By 
        Hi
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginLeft": 8,
            "marginTop": 3,
          }
        }
      >
        <Text />
        <Text
          style={
            Object {
              "marginLeft": 10,
            }
          }
        >
          Concordia
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginLeft": 8,
            "marginTop": 3,
          }
        }
      >
        <Text />
        <Text
          style={
            Object {
              "marginLeft": 10,
            }
          }
        >
          Jun 21, 2023
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginLeft": 8,
            "marginTop": 3,
          }
        }
      >
        <Text />
        <Text
          style={
            Object {
              "marginLeft": 10,
            }
          }
        >
          08:13 pm - 08:13 pm
        </Text>
      </View>
      <Text
        style={
          Object {
            "color": "#100101",
            "fontFamily": "Helvetica Neue",
            "fontSize": 17,
            "marginBottom": "2%",
            "marginLeft": "1%",
            "marginTop": "5%",
          }
        }
      >
        Description
      </Text>
      <Text
        style={
          Object {
            "alignContent": "flex-start",
            "alignSelf": "flex-start",
            "fontFamily": "Helvetica Neue",
            "fontSize": 14,
            "marginLeft": "1%",
          }
        }
      >
        Fun
      </Text>
      <Text
        style={
          Object {
            "color": "#100101",
            "fontFamily": "Helvetica Neue",
            "fontSize": 17,
            "marginBottom": "2%",
            "marginLeft": "1%",
            "marginTop": "5%",
          }
        }
      >
        Location
      </Text>
      <View
        style={
          Object {
            "borderRadius": 20,
            "overflow": "hidden",
          }
        }
      >
        <MapView
          initialRegion={
            Object {
              "latitude": 35.6762,
              "latitudeDelta": 0.01,
              "longitude": 139.6503,
              "longitudeDelta": 0.01,
            }
          }
          style={
            Object {
              "borderRadius": 20,
              "height": 333.5,
              "overflow": "hidden",
              "width": 750,
            }
          }
        >
          <Marker
            coordinate={
              Object {
                "latitude": 35.6762,
                "latitudeDelta": 0.01,
                "longitude": 139.6503,
                "longitudeDelta": 0.01,
              }
            }
            pinColor="#00c0a4"
          />
        </MapView>
      </View>
      <View
        style={
          Object {
            "alignSelf": "center",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "justifyContent": "flex-start",
            "marginTop": 10,
            "width": "95%",
          }
        }
      >
        <Text />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
